src/expr/src/vector_op/agg/approx_count_distinct.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+
+use risingwave_common::array::*;
+use risingwave_common::error::{ErrorCode, Result};
+use risingwave_common::types::*;
+
+use crate::vector_op::agg::aggregator::Aggregator;
+use crate::vector_op::agg::general_sorted_grouper::EqGroups;
+
+const INDEX_BITS: u8 = 10; // number of bits used for finding the index of each 64-bit hash
+const INDICES: usize = 1 << INDEX_BITS; // number of indices available
+const COUNT_BITS: u8 = 64 - INDEX_BITS; // number of non-index bits in each 64-bit hash
+
+pub struct ApproxCountDistinct {
+    return_type: DataType,
+    input_col_idx: usize,
+    registers: [u8; INDICES],
+}
+
+impl ApproxCountDistinct {
+    pub fn new(return_type: DataType, input_col_idx: usize) -> Self {
+        Self {
+            return_type,
+            input_col_idx,
+            registers: [0; INDICES],
+        }
+    }
+
+    /// Adds the count of the datum's hash into the register, if it is greater than the existing
+    /// count at the register
+    fn add_datum(&mut self, datum_ref: DatumRef) {
+        if datum_ref.is_none() {
+            return;
+        }
+
+        let scalar_impl = datum_ref.unwrap().into_scalar_impl();
+        let hash = self.get_hash(scalar_impl);
+
+        let index = (hash as usize) & (INDICES - 1); // Index is based on last few bits
+        let count = self.count_hash(hash);
+
+        if count > self.registers[index] {
+            self.registers[index] = count;
+        }
+    }
+
+    /// Calculate the hash of the `scalar_impl` using Rust's default hasher
+    /// Perhaps a different hash like Murmur2 could be used instead for optimization?
+    fn get_hash(&self, scalar_impl: ScalarImpl) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        scalar_impl.hash(&mut hasher);
+        hasher.finish()
+    }
+
+    /// Counts the number of trailing zeroes plus 1 in the non-index bits of the hash
+    fn count_hash(&self, mut hash: u64) -> u8 {
+        let mut count = 1;
+
+        hash >>= INDEX_BITS; // Ignore bits used as index for the hash
+        hash |= 1 << COUNT_BITS; // To allow hash to terminate if it is all 0s
+
+        while hash & 1 == 0 {
+            count += 1;
+            hash >>= 1;
+        }
```

Comment:
X86 supports a bit-scan instruction to do this efficiently, which can be called by `__builtin_ctz` in C++. Is it possible to use that in Rust?